File: packages/posthog/ios/Plugin/Posthog.swift
import Foundation
import PostHog

@objc public class Posthog: NSObject {
    init(config: PosthogConfig) {
        super.init()
        if let apiKey = config.apiKey {
            self.setup(
                apiKey: apiKey,
                apiHost: config.apiHost,
                enableSessionReplay: config.enableSessionReplay,
                captureApplicationLifecycleEvents: config.captureApplicationLifecycleEvents,
                autoCaptureExceptions: config.autoCaptureExceptions,
                sessionReplayConfig: config.sessionReplayConfig
            )

            // Start session recording if configured
            if config.enableSessionReplay {
                self.startSessionRecording()
            }
        }
    }

    @objc public func alias(_ options: AliasOptions) {
        let alias = options.getAlias()

        PostHogSDK.shared.alias(alias)
    }

    @objc public func capture(_ options: CaptureOptions) {
        let event = options.getEvent()
        let properties = options.getProperties()

        PostHogSDK.shared.capture(event, properties: properties)
    }

    @objc public func captureException(_ options: CaptureExceptionOptions) {
        // Emit the `$exception` event directly instead of handing the SDK an `Error`.
        // The `Error` path re-derives the type from the error domain and discards the
        // JavaScript stack in favour of a native one, while we want the JavaScript name
        // and stack. This mirrors the `$exception_list` payload that posthog-js produces.
        let name = options.getName() ?? ""
        var exception: [String: Any] = [
            "type": name.isEmpty ? "Error" : name,
            "value": options.getMessage(),
            "mechanism": [
                "type": "generic",
                "handled": true,
                "synthetic": false
            ]
        ]

        let frames = Posthog.createFrames(options.getStacktrace())
        if !frames.isEmpty {
            exception["stacktrace"] = [
                "type": "raw",
                "frames": frames
            ]
        }

        var properties: [String: Any] = [:]
        options.getProperties()?.forEach { properties[$0.key] = $0.value }
        properties["$exception_level"] = "error"
        properties["$exception_list"] = [exception]

        PostHogSDK.shared.capture("$exception", properties: properties)
    }

    private static func createFrames(_ stacktrace: [StackFrame]?) -> [[String: Any]] {
        guard let stacktrace = stacktrace, !stacktrace.isEmpty else {
            return []
        }
        var frames: [[String: Any]] = []
        for frame in stacktrace {
            var map: [String: Any] = [
                "platform": "web:javascript",
                "function": frame.getFunctionName() ?? "?",
                "in_app": true
            ]
            if let fileName = frame.getFileName() {
                map["filename"] = fileName
            }
            if let lineNumber = frame.getLineNumber() {
                map["lineno"] = lineNumber
            }
            if let columnNumber = frame.getColumnNumber() {
                map["colno"] = columnNumber
            }
            frames.append(map)
        }
        // PostHog stores frames bottom-up (oldest call first), while the JavaScript
        // stack trace is top-down (most recent call first), so reverse them.
        return frames.reversed()
    }

    @objc public func flush() {
        PostHogSDK.shared.flush()
    }

    @objc public func getDistinctId() -> GetDistinctIdResult {
        let distinctId = PostHogSDK.shared.getDistinctId()
        return GetDistinctIdResult(distinctId: distinctId)
    }

    @objc public func getFeatureFlag(_ options: GetFeatureFlagOptions) -> GetFeatureFlagResult {
        let key = options.getKey()

        let value = PostHogSDK.shared.getFeatureFlag(key)
        return GetFeatureFlagResult(value: value)
    }

    @objc public func getFeatureFlagPayload(_ options: GetFeatureFlagPayloadOptions) -> GetFeatureFlagPayloadResult {
        let key = options.getKey()

        let value = PostHogSDK.shared.getFeatureFlagPayload(key)
        return GetFeatureFlagPayloadResult(value: value)
    }

    @objc public func group(_ options: GroupOptions) {
        let type = options.getType()
        let key = options.getKey()
        let groupProperties = options.getGroupProperties()

        PostHogSDK.shared.group(type: type, key: key, groupProperties: groupProperties)
    }

    @objc public func identify(_ options: IdentifyOptions) {
        let distinctId = options.getDistinctId()
        let userProperties = options.getUserProperties()

        PostHogSDK.shared.identify(distinctId, userProperties: userProperties)
    }

    @objc public func isFeatureEnabled(_ options: IsFeatureEnabledOptions) -> IsFeatureEnabledResult {
        let key = options.getKey()

        let isEnabled = PostHogSDK.shared.isFeatureEnabled(key)
        return IsFeatureEnabledResult(enabled: isEnabled)
    }

    @objc public func isOptOut() -> Bool {
        return PostHogSDK.shared.isOptOut()
    }

    @objc public func optIn() {
        PostHogSDK.shared.optIn()
    }

    @objc public func optOut() {
        PostHogSDK.shared.optOut()
    }

    @objc public func register(_ options: RegisterOptions) {
        let key = options.getKey()
        let value = options.getValue()

        let properties = [key: value]

        PostHogSDK.shared.register(properties)
    }

    @objc public func reloadFeatureFlags() {
        PostHogSDK.shared.reloadFeatureFlags()
    }

    @objc public func reset() {
        PostHogSDK.shared.reset()
    }

    @objc public func screen(_ options: ScreenOptions) {
        let screenTitle = options.getScreenTitle()
        let properties = options.getProperties()

        PostHogSDK.shared.screen(screenTitle, properties: properties)
    }

    @objc public func setup(_ options: SetupOptions) {
        let apiKey = options.getApiKey()
        let apiHost = options.getApiHost()
        let enableSessionReplay = options.getEnableSessionReplay()
        let optOut = options.getOptOut()
        let captureApplicationLifecycleEvents = options.getCaptureApplicationLifecycleEvents()
        let autoCaptureExceptions = options.getAutoCaptureExceptions()
        let sessionReplayConfig = options.getSessionReplayConfig()

        setup(
            apiKey: apiKey,
            apiHost: apiHost,
            enableSessionReplay: enableSessionReplay,
            optOut: optOut,
            captureApplicationLifecycleEvents: captureApplicationLifecycleEvents,
            autoCaptureExceptions: autoCaptureExceptions,
            sessionReplayConfig: sessionReplayConfig
        )
    }

    @objc public func startSessionRecording() {
        PostHogSDK.shared.startSessionRecording()
    }

    @objc public func stopSessionRecording() {
        PostHogSDK.shared.stopSessionRecording()
    }

    private func setup(apiKey: String, apiHost: String, enableSessionReplay: Bool = false, optOut: Bool = false, captureApplicationLifecycleEvents: Bool = true, autoCaptureExceptions: Bool = false, sessionReplayConfig: SessionReplayOptions? = nil) {
        let config = PostHogConfig(apiKey: apiKey, host: apiHost)
        config.captureScreenViews = false
        config.optOut = optOut
        config.captureApplicationLifecycleEvents = captureApplicationLifecycleEvents
        config.errorTrackingConfig.autoCapture = autoCaptureExceptions
        config.sessionReplay = enableSessionReplay
        config.sessionReplayConfig.screenshotMode = sessionReplayConfig?.getScreenshotMode() ?? false
        config.sessionReplayConfig.maskAllImages = sessionReplayConfig?.getMaskAllImages() ?? true
        config.sessionReplayConfig.maskAllTextInputs = sessionReplayConfig?.getMaskAllTextInputs() ?? true
        config.sessionReplayConfig.maskAllSandboxedViews = sessionReplayConfig?.getMaskAllSandboxedViews() ?? true
        config.sessionReplayConfig.captureNetworkTelemetry = sessionReplayConfig?.getCaptureNetworkTelemetry() ?? true
        config.sessionReplayConfig.debouncerDelay = sessionReplayConfig?.getDebouncerDelay() ?? 1.0

        PostHogSDK.shared.setup(config)
    }

    @objc public func unregister(_ options: UnregisterOptions) {
        let key = options.getKey()

        PostHogSDK.shared.unregister(key)
    }
}

File: packages/posthog/android/src/main/java/io/capawesome/capacitorjs/plugins/posthog/Posthog.kt
package io.capawesome.capacitorjs.plugins.posthog

import com.posthog.PostHog
import com.posthog.android.PostHogAndroid
import io.capawesome.capacitorjs.plugins.posthog.classes.options.CaptureExceptionOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.CaptureOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.IdentifyOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.RegisterOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.ScreenOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.SetupOptions

import com.posthog.android.PostHogAndroidConfig
import io.capawesome.capacitorjs.plugins.posthog.classes.options.AliasOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.GetFeatureFlagOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.GetFeatureFlagPayloadOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.GroupOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.IsFeatureEnabledOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.SessionReplayOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.options.StackFrame
import io.capawesome.capacitorjs.plugins.posthog.classes.options.UnregisterOptions
import io.capawesome.capacitorjs.plugins.posthog.classes.results.GetDistinctIdResult
import io.capawesome.capacitorjs.plugins.posthog.classes.results.GetFeatureFlagPayloadResult
import io.capawesome.capacitorjs.plugins.posthog.classes.results.GetFeatureFlagResult
import io.capawesome.capacitorjs.plugins.posthog.classes.results.IsFeatureEnabledResult

class Posthog(private val config: PosthogConfig, private val plugin: PosthogPlugin) {

    init {
        val apiKey = config.getApiKey()
        if (apiKey != null) {
            setup(
                apiKey,
                config.getApiHost(),
                config.getEnableSessionReplay(),
                false,
                config.getCaptureApplicationLifecycleEvents(),
                config.getAutoCaptureExceptions(),
                config.getSessionReplayConfig()
            )
        }
    }

    fun alias(options: AliasOptions) {
        val alias = options.alias
        com.posthog.PostHog.alias(alias = alias)
    }

    fun capture(options: CaptureOptions) {
        val event = options.event
        val properties = options.properties

        com.posthog.PostHog.capture(event = event, properties = properties)
    }

    fun captureException(options: CaptureExceptionOptions) {
        // Emit the `$exception` event directly instead of handing the SDK a Throwable.
        // The Throwable path derives the type from the throwable's class name (dropping the
        // caller's name) and synthesizes a native stack, while we want the JavaScript name
        // and stack. This mirrors the `$exception_list` payload that posthog-js produces.
        val name = options.name?.takeIf { it.isNotEmpty() } ?: "Error"
        val exception = mutableMapOf<String, Any>(
            "type" to name,
            "value" to options.message,
            "mechanism" to mapOf(
                "type" to "generic",
                "handled" to true,
                "synthetic" to false
            )
        )

        val frames = createFrames(options.stacktrace)
        if (frames.isNotEmpty()) {
            exception["stacktrace"] = mapOf(
                "type" to "raw",
                "frames" to frames
            )
        }

        val properties = options.properties?.toMutableMap() ?: mutableMapOf()
        properties["\$exception_level"] = "error"
        properties["\$exception_list"] = listOf(exception)

        com.posthog.PostHog.capture(event = "\$exception", properties = properties)
    }

    fun flush() {
        com.posthog.PostHog.flush()
    }

    fun getDistinctId(): GetDistinctIdResult {
        val distinctId = com.posthog.PostHog.distinctId()
        return GetDistinctIdResult(distinctId)
    }

    fun getFeatureFlag(options: GetFeatureFlagOptions): GetFeatureFlagResult {
        val key = options.key

        val value = com.posthog.PostHog.getFeatureFlag(key = key)
        return GetFeatureFlagResult(value)
    }

    fun getFeatureFlagPayload(options: GetFeatureFlagPayloadOptions): GetFeatureFlagPayloadResult {
        val key = options.key

        val value = com.posthog.PostHog.getFeatureFlagPayload(key = key)
        return GetFeatureFlagPayloadResult(value)
    }

    fun group(options: GroupOptions) {
        val type = options.type
        val key = options.key
        val groupProperties = options.groupProperties

        com.posthog.PostHog.group(type = type, key = key, groupProperties = groupProperties)
    }

    fun identify(options: IdentifyOptions) {
        val distinctId = options.distinctId
        val userProperties = options.userProperties

        com.posthog.PostHog.identify(distinctId = distinctId, userProperties = userProperties)
    }

    fun isFeatureEnabled(options: IsFeatureEnabledOptions): IsFeatureEnabledResult {
        val key = options.key

        val isEnabled = com.posthog.PostHog.isFeatureEnabled(key = key, defaultValue = false)
        return IsFeatureEnabledResult(isEnabled)
    }

    fun isOptOut(): Boolean {
        return com.posthog.PostHog.isOptOut()
    }

    fun optIn() {
        com.posthog.PostHog.optIn()
    }

    fun optOut() {
        com.posthog.PostHog.optOut()
    }

    fun register(options: RegisterOptions) {
        val key = options.key
        val value = options.value

        com.posthog.PostHog.register(key = key, value = value)
    }

    fun reloadFeatureFlags() {
        com.posthog.PostHog.reloadFeatureFlags()
    }

    fun reset() {
        com.posthog.PostHog.reset()
    }

    fun screen(options: ScreenOptions) {
        val name = options.screenTitle
        val properties = options.properties

        com.posthog.PostHog.screen(screenTitle = name, properties = properties)
    }

    fun setup(options: SetupOptions) {
        val apiKey = options.apiKey
        val apiHost = options.apiHost
        val enableSessionReplay = options.enableSessionReplay
        val optOut = options.optOut
        val captureApplicationLifecycleEvents = options.captureApplicationLifecycleEvents
        val autoCaptureExceptions = options.autoCaptureExceptions
        val sessionReplayConfig = options.sessionReplayConfig

        setup(
            apiKey,
            apiHost,
            enableSessionReplay,
            optOut,
            captureApplicationLifecycleEvents,
            autoCaptureExceptions,
            sessionReplayConfig
        )
    }

    fun unregister(options: UnregisterOptions) {
        val key = options.key

        com.posthog.PostHog.unregister(key = key)
    }

    private fun createFrames(stacktrace: List<StackFrame>?): List<Map<String, Any>> {
        if (stacktrace.isNullOrEmpty()) {
            return emptyList()
        }
        val frames = stacktrace.map { frame ->
            val map = mutableMapOf<String, Any>(
                "platform" to "web:javascript",
                "function" to (frame.functionName ?: "?"),
                "in_app" to true
            )
            frame.fileName?.let { map["filename"] = it }
            frame.lineNumber?.let { map["lineno"] = it }
            frame.columnNumber?.let { map["colno"] = it }
            map
        }
        // PostHog stores frames bottom-up (oldest call first), while the JavaScript
        // stack trace is top-down (most recent call first), so reverse them.
        return frames.asReversed()
    }

    private fun setup(
        apiKey: String,
        apiHost: String,
        enableSessionReplay: Boolean = false,
        optOut: Boolean = false,
        captureApplicationLifecycleEvents: Boolean = true,
        autoCaptureExceptions: Boolean = false,
        sessionReplayConfig: SessionReplayOptions? = null
    ) {
        val posthogConfig = PostHogAndroidConfig(
            apiKey = apiKey,
            host = apiHost
        )
        posthogConfig.captureScreenViews = false
        posthogConfig.captureApplicationLifecycleEvents = captureApplicationLifecycleEvents
        posthogConfig.optOut = optOut
        posthogConfig.sessionReplay = enableSessionReplay
        posthogConfig.errorTrackingConfig.autoCapture = autoCaptureExceptions

        // Configure session replay options if provided
        sessionReplayConfig?.let { replayConfig ->
            // Screenshot mode configuration
            posthogConfig.sessionReplayConfig.screenshot = replayConfig.getScreenshotMode() ?: false

            // Text input masking
            posthogConfig.sessionReplayConfig.maskAllTextInputs = replayConfig.getMaskAllTextInputs() ?: true

            // Image masking
            posthogConfig.sessionReplayConfig.maskAllImages = replayConfig.getMaskAllImages() ?: true

            // Network telemetry capture
            posthogConfig.sessionReplayConfig.captureLogcat = replayConfig.getCaptureNetworkTelemetry() ?: true

            // Debounce delay for performance (convert seconds to milliseconds)
            val debouncerDelaySeconds = replayConfig.getDebouncerDelay() ?: 1.0
            posthogConfig.sessionReplayConfig.throttleDelayMs = (debouncerDelaySeconds * 1000).toLong()
        }

        PostHogAndroid.setup(plugin.context, posthogConfig)
    }
}
